### PR TITLE
Added e2e framweork and a simple test for HA master.

### PR DESCRIPTION
**Pull request by jszczepkowski**
Added e2e framweork for testing HA master. Added one simple e2e test for HA master that first grows and then shrinks GCE cluster.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32860)

### CODE REVIEW

File: test/e2e/ha_master.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	"strconv"
+)
+
+func addMasterReplica() error {
+	framework.Logf(fmt.Sprintf("Adding a new master replica:"))
+	v, _, err := framework.RunCmd(path.Join(framework.TestContext.RepoRoot, "hack/e2e-internal/e2e-add-master.sh"))
+	if err != nil {
+		return err
+	}
+	framework.Logf("%q", v)
```
Comment (fgrzadkowski) [inline test/e2e/ha_master.go]: Shouldn't we also print output in case of error?

File: test/e2e/ha_master.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	return nil
+}
+
+func removeMasterReplica() error {
+	framework.Logf(fmt.Sprintf("Removing an existing master replica:"))
+	v, _, err := framework.RunCmd(path.Join(framework.TestContext.RepoRoot, "hack/e2e-internal/e2e-remove-master.sh"))
+	if err != nil {
+		return err
+	}
+	framework.Logf("%q", v)
```
Comment (fgrzadkowski) [inline test/e2e/ha_master.go]: Ditto.

File: test/e2e/ha_master.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+func createNewRC(c *client.Client, ns string, name string) {
+	_, err := newRCByName(c, ns, name, 1)
+	framework.ExpectNoError(err)
+}
+
+func verifyNumberOfMasterReplicas(expected int) {
+	output, err := exec.Command("gcloud", "compute", "instances", "list",
+		"--project="+framework.TestContext.CloudConfig.ProjectID,
+		"--zone="+framework.TestContext.CloudConfig.Zone,
```
Comment (fgrzadkowski) [inline test/e2e/ha_master.go]: I don't think we should limit this query to a single zone.

File: test/e2e/ha_master.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+		"--regexp="+framework.TestContext.CloudConfig.MasterName + "(-...)?").CombinedOutput()
+	framework.ExpectNoError(err)
+	re := regexp.MustCompile(".*RUNNING")
```
Comment (fgrzadkowski) [inline test/e2e/ha_master.go]: I think you just use gcloud filtering:

`--filter=status=RUNNING`

File: test/e2e/ha_master.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+		case AddReplica:
+			framework.ExpectNoError(addMasterReplica())
+			additionalReplicas++
+		case RemoveReplica:
+			framework.ExpectNoError(removeMasterReplica())
+			additionalReplicas--
+		}
+		verifyNumberOfMasterReplicas(additionalReplicas + 1)
+
+		rcName := "ha-master-" + strconv.Itoa(len(existingRCs))
```
Comment (fgrzadkowski) [inline test/e2e/ha_master.go]: Is this step to verify that apiserver works in HA mode? If so, can you please add a comment?

File: test/e2e/ha_master.go (hunk 1 above)
Comment (jszczepkowski) [inline test/e2e/ha_master.go]: done

File: test/e2e/ha_master.go (hunk 2 above)
Comment (jszczepkowski) [inline test/e2e/ha_master.go]: done

File: test/e2e/ha_master.go (hunk 3 above)
Comment (jszczepkowski) [inline test/e2e/ha_master.go]: Currently, e2e tests run in a single zone: framework.TestContext.CloudConfig.Zone. So, there is no masters in a different zone. If we need multi-zone e2e test in future for ha-master (most probably: we will need), I will add handling additional zone(s) to test Context and in all required places (including here).

File: test/e2e/ha_master.go (hunk 4 above)
Comment (jszczepkowski) [inline test/e2e/ha_master.go]: This regexp has two purposes:
- include only RUNNING nodes, 
- exclude table description (couldn't find gcloud flag for this).
Even if I add the filtering, I will still need kind of a regexp here.

File: test/e2e/ha_master.go (hunk 5 above)
Comment (jszczepkowski) [inline test/e2e/ha_master.go]: done

File: test/e2e/ha_master.go (hunk 4 above)
Comment (fgrzadkowski) [inline test/e2e/ha_master.go]: Still doable:

```
$ gcloud compute instances list --regexp=".*master.*" --filter=status=RUNNING --format=[no-heading]
kubernetes-master  us-central1-b  n1-standard-1    10.128.0.8  146.148.92.107  RUNNING
```

See [here](https://cloud.google.com/sdk/gcloud/reference/topic/formats) for more formats.

Comment (jszczepkowski) [inline test/e2e/ha_master.go]: Thanks!!! Done

**Comment by jszczepkowski** [comment]
@fgrzadkowski 
Comments applied, PTAL

File: test/e2e/ha_master.go [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+	f := framework.NewDefaultFramework("ha-master")
+	var c *client.Client
+	var ns string
+	var additionalReplicas int
+	var existingRCs []string
+
+	BeforeEach(func() {
+		c = f.Client
+		ns = f.Namespace.Name
+		verifyNumberOfMasterReplicas(1)
```
Comment (dshulyak) [inline test/e2e/ha_master.go]: should be skipped if provider != gce

Comment (jszczepkowski) [inline test/e2e/ha_master.go]: Thanks for noticing this! Fixed

**Comment by fgrzadkowski** [comment]
Apart from the Jenkins failures it LGTM.

**Comment by jszczepkowski** [comment]
@k8s-bot test this issue: #33184

**Comment by jszczepkowski** [comment]
Part of https://github.com/kubernetes/features/issues/48

**Pull request merged**